### PR TITLE
Add functionality to append rounds to existing config files

### PR DESCRIPTION
**Pull request by annakrystalli**
This PR implements the fixes described in #42 , specifically it:

- * Adds the `as_config()` function to convert a list representation of a `tasks.json` config file to a `<config>` class object. Useful when wanting to programmatically manipulate the contents of a `tasks.json` config file.
* Adds the `append_round()` function to append one or more `<round>` class objects to the `rounds` property of a `<config>` class object. 

I've also set the package version to the development version

### CODE REVIEW

**Review by bsweger — APPROVED** [review]
Thanks for addressing this so quickly!

I gave the new features a spin on a simplified version of the variant-nowcast-hub script we've been working on: https://github.com/reichlab/variant-nowcast-hub/blob/bsweger/test-round-append/src/append_round_test.R

It worked great, and will definitely simplify the variant nowcast code.

One observation (which may be what we want, not sure). The old tasks.json file in the main branch of variant-nowcast-hub still points to the schema in the old `infectious-disease-modeling-hub` org name, and I had to change it `hubverse-org` to avoid this error:

```
Error in `append_round()` at src/append_round_test.R:67:1:
✖ Schema version mismatch between `config` and round.
! Must be the same.
• `config`:
  "https://raw.githubusercontent.com/Infectious-Disease-Modeling-Hubs/schemas/main/v3.0.1/tasks-schema.json"
• round: "https://raw.githubusercontent.com/hubverse-org/schemas/main/v3.0.1/tasks-schema.json"
```

Which seems fine, since we want people to use the new org name anyway, but I wasn't sure of your intent.

File: tests/testthat/test-append_round.R [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+  # Adding two round works
+  add_2_rounds <- append_round(
+    config = config, round,
+    create_new_round("2024-09-25")
+  )
+
+  expect_length(add_2_rounds$rounds, 3)
+  expect_s3_class(add_2_rounds, c("config", "list"))
+})
+
```
Comment (zkamvar) [inline tests/testthat/test-append_round.R:19]: Here's the test for the version of the function that updates the schema URL.

```suggestion

test_that("append_round allows for and updates old schema url", {
  config <- hubUtils::read_config_file(
    config_path = test_path("testdata/tasks-append.json")
  ) |> as_config()
  round <- create_new_round()
  ver <- hubUtils::extract_schema_version(attr(config, "schema_id"))
  cfg <- basename(attr(config, "schema_id"))
  ver <- file.path(ver, cfg)
  prefix <- "https://raw.githubusercontent.com/"
  owner <- "Infectious-Disease-Modeling-Hubs" 
  suffix <- "/schemas/main/"
  attr(config, "schema_id") <- paste0(prefix, owner, suffix, ver)
 
  # Adding one round works
  add_round <- append_round(config, round)
  expect_length(add_round$rounds, 2)
  expect_s3_class(add_round, c("config", "list"))
  expected <- paste(prefix, "hubverse-org", suffix, ver)
  expect_equal(attr(add_round, "schema_id"), expected)
})

```

**Review by zkamvar — APPROVED** [review]
This looks fantastic! 

I have some suggestions that might help solve the thing Becky ran into and prevent errors downstream (I think). Nothing is a blocker. Great job!

File: R/append_round.R [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+#'   system.file("tasks.json", package = "hubAdmin")
+#' ) |> as_config()
+#' append_round(config, new_round)
+append_round <- function(config, ...) {
+  checkmate::assert_class(config, "config")
+
+  rounds <- create_rounds(...)
+
+  config_schema_version <- attr(config, "schema_id")
+  round_schema_version <- attr(rounds, "schema_id")
```
[the remark this answers is not in the document]
Comment (annakrystalli) [reply R/append_round.R:130]: Although we still support the old URL prefix for back-compatibility in some situations, it is for all intents and purposes considered superseded so I would not want to accept it as a match to the newer and correct URL. In fact this is a good opportunity to remind users to  update that field which is what happened for @bsweger (i.e. I feel the correct solution was indeed for the URL to be updated from the superseded prefix, not for them to be considered a match).

Having said that I do like the solution you propose below of automatically updating any superseded prefixes!

File: R/as_config.R [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+    schema_url,
+    fixed = FALSE
+  )
+  if (!check_prefix) {
+    cli::cli_abort(
+      c("x" = "Invalid {.field {property_name}} property. Should start with:
+              {.val https://raw.githubusercontent.com/hubverse-org/schemas/main/}"),
+      call = call
+    )
+  }
```
[the remark this answers is not in the document]
Comment (annakrystalli) [reply R/as_config.R:86]: I like this suggestion! I wouldn't want to change anything without issuing a message about it though

**Comment by annakrystalli** [comment]
@bsweger  & @zkamvar, thank you both for your reviews! 

@zkamvar I've responded to your suggestions individually but to summarise, while I'm not keen on accepting old schema URL prefixes as matches to correct current ones, I very much like the idea of catching them and correcting them automatically. I do feel we should be issuing a message when we do that though. 

So there's some minor edits required in your proposed changes in `validate_schema_url_prefix()` and the tests you proposed. Would you like to directly edit your proposals and then commit them or would you like me to commit them and make the changes myself?

File: R/as_config.R (hunk 3 above)
[the remark this answers is not in the document]
Comment (zkamvar) [reply R/as_config.R:86]: @annakrystalli I updated the suggestion to include an alert.

**Comment by zkamvar** [comment]
> So there's some minor edits required in your proposed changes in `validate_schema_url_prefix()` and the tests you proposed. Would you like to directly edit your proposals and then commit them or would you like me to commit them and make the changes myself?

I've edited my comments directly

**Comment by annakrystalli** [comment]
Hmmm looking at it more I think the correction should be part of `as_config()` instead of `append_config()`? what do you think @zkamvar ?

**Comment by zkamvar** [comment]
> Hmmm looking at it more I think the correction should be part of `as_config()` instead of `append_config()`? what do you think @zkamvar ?

I think that's a great idea! And actually... looking at the code now, it's not going to work as it is because the modified URL goes nowhere 😅.

I'll make some edits to account for this.

File: tests/testthat/test-as_config.R [hunk 4]
```diff
@@ -0,0 +1,34 @@
+test_that("as_config succeeds with valid config", {
+  skip_if_offline()
+  config_tasks <- hubUtils::read_config_file(
+    config_path = test_path("testdata/tasks-append.json")
+  )
+  expect_snapshot(as_config(config_tasks))
+})
+
```
Comment (zkamvar) [inline tests/testthat/test-as_config.R:8]:
```suggestion
test_that("as_config succeeds with valid config", {
  skip_if_offline()
  config_tasks <- hubUtils::read_config_file(
    config_path = test_path("testdata/tasks-append.json")
  )
  expect_snapshot(as_config(config_tasks))
})

test_that("as_config modifies outdated URLs", {
  skip_if_offline()
  orig <- config_tasks <- hubUtils::read_config_file(
    config_path = test_path("testdata/tasks-append.json")
  )
  config_tasks$schema_version <- sub("hubverse-org", 
    "Infectious-Disease-Modelling-Hubs",
    config_tasks$schema_version
  )
  # a message will be issued that the URL is updated
  expect_message(res <- as_config(config_tasks), 'superseded URL')
  # the resulting should be equal to the original 
  expect_equal(orig$schema_version, res$schema_version)
})

```

File: tests/testthat/test-append_round.R (hunk 1 above)
Comment (zkamvar) [reply tests/testthat/test-append_round.R:19]: resolving this in favor of Anna's suggestion

**Comment by annakrystalli** [comment]
Thanks for the additions @zkamvar ! I think I committed everything correctly...I guess we'll find out shortly! 😁

**Comment by annakrystalli** [comment]
Hmmm, I thought the lint would fal but looks like more. I'll pull down and check it

**Comment by annakrystalli** [comment]
OK trying again. There was a tiny typo in the tests but I think all's good now!

**Comment by annakrystalli** [comment]
Woohoo! Passed! Merging. Thanks again @zkamvar and @bsweger

**Pull request merged**